### ISSUE TITLE
Prompt caching causing unstable generation

### ISSUE THREAD

**Issue by cbowdon** [madroidmaq/mlx-omni-server#57]
Hi! This is such a great project, thanks for making it.

I've noticed that the log probs for repeat requests are not stable even when the sampling params would make them be stable. It is because of prompt caching: when prompt caching is disabled, they become stable. 

As far as I can tell, the contents of the cache do not match the omitted portion of the prompt. Trimming the cache more than necessary resolves the issue, but I'm not yet sure why. I think the final <|im_end|> token is a part of the cause, as this is in the prompt cache but not the stored tokens.

**Comment by cbowdon** [comment]
I'm 99% sure it's something to do with chat templates. I adapted some prompt caching tests from the mlx-lm repo to reflect the usage in mlx-omni-server, and look.

This test passes:

```python
def test_trim_cache_with_generate(MODEL=QWEN3):
    model, tokenizer = load(MODEL)
    prompt = tokenizer.encode("this is a prompt", return_tensors="mlx")[0]

    prompt_cache = make_prompt_cache(model)

    # Generate n tokens
    n = 8
    results = zip(range(n), generate_step(prompt, model, prompt_cache=prompt_cache))
    toks, all_logits = zip(*(r[1] for r in results))

    # To get back to the cache to some offset from the end of the last prompt
    # trim by n + offset tokens
    offset = 3
    trim_prompt_cache(prompt_cache, n + offset)

    # Generate the same thing again
    results = zip(
        range(n), generate_step(prompt[-offset:], model, prompt_cache=prompt_cache)
    )
    second_toks, second_all_logits = zip(*(r[1] for r in results))
    assert toks == second_toks
    assert all(mx.allclose(l, l2) for l, l2 in zip(all_logits, second_all_logits))
```

But when I put in a templated prompt, the logits become unstable!

```python
def test_chat_template_trim_cache_with_generate(MODEL=QWEN3):
    model, tokenizer = load(MODEL)
    templated_prompt = """
<|im_start>user
this is a prompt<|im_end|>
<|im_start|>assistant
<think>

</think>
""".strip()
    prompt = tokenizer._tokenizer.encode(templated_prompt, return_tensors="mlx")[0]

    prompt_cache = make_prompt_cache(model)

    # Generate n tokens
    n = 8
    results = zip(range(n), generate_step(prompt, model, prompt_cache=prompt_cache))
    toks, all_logits = zip(*(r[1] for r in results))

    # To get back to the cache to some offset from the end of the last prompt
    # trim by n + offset tokens
    offset = 3
    trim_prompt_cache(prompt_cache, n + offset)

    # Generate the same thing again
    results = zip(
        range(n), generate_step(prompt[-offset:], model, prompt_cache=prompt_cache)
    )
    second_toks, second_all_logits = zip(*(r[1] for r in results))
    assert toks == second_toks
    # THIS ONE FAILS
    assert all(mx.allclose(l, l2) for l, l2 in zip(all_logits, second_all_logits))
```

Something somewhere in `mlx` or in the model treats special template tokens differently? Maybe it doesn't cache them? I don't know. Hope someone smarter than me can help.

**Comment by cbowdon** [comment]
No bug apparently, just what you should expect for prompt caching.
